File: cxx/test/test_unionfind.cpp
#include <gtest/gtest.h>
#include "../spanning/unionfind.hpp"


TEST(UnionFind, mini_test_2)
{
    UnionFind<int> u (2);
    ASSERT_EQ(u.find(1), 1);
    ASSERT_EQ(u.find(0), 0);

    u.unite(u.find(1), u.find(0));
    ASSERT_EQ(u.find(0), u.find(1));
    ASSERT_EQ(u.find(1), u.find(0));
}


TEST(UnionFind, mini_test_3)
{
    UnionFind<int> u (3);
    ASSERT_EQ(u.find(2), 2);
    ASSERT_EQ(u.find(1), 1);
    ASSERT_EQ(u.find(0), 0);

    u.unite(u.find(2), u.find(0));
    EXPECT_EQ(u.find(0), u.find(2));
    EXPECT_EQ(u.find(2), u.find(0));

    EXPECT_NE(u.find(2), u.find(1));
    EXPECT_NE(u.find(1), u.find(0));

    u.unite(u.find(1), u.find(0));
    EXPECT_EQ(u.find(2), u.find(1));
    EXPECT_EQ(u.find(1), u.find(0));
    EXPECT_EQ(u.find(0), u.find(1));
}


TEST(UnionFind, mini_test_4)
{
    UnionFind<int> u (4);
    ASSERT_EQ(u.find(3), 3);
    ASSERT_EQ(u.find(2), 2);
    ASSERT_EQ(u.find(1), 1);
    ASSERT_EQ(u.find(0), 0);

    u.unite(u.find(3), u.find(0));
    EXPECT_EQ(u.find(3), u.find(0));
    EXPECT_EQ(u.find(3), u.find(0));
    EXPECT_EQ(u.find(0), u.find(0));
    EXPECT_EQ(u.find(0), u.find(3));

    EXPECT_NE(u.find(3), u.find(2));
    EXPECT_NE(u.find(2), u.find(3));
    EXPECT_NE(u.find(2), u.find(0));
    EXPECT_NE(u.find(2), u.find(0));

    u.unite(u.find(3), u.find(2));
    EXPECT_EQ(u.find(3), u.find(2));
    EXPECT_EQ(u.find(2), u.find(3));
    EXPECT_EQ(u.find(2), u.find(0));
    EXPECT_EQ(u.find(2), u.find(0));
    EXPECT_EQ(u.find(0), u.find(2));
}


TEST(UnionFind, mini_test_4a)
{
    UnionFind<int> u (4);
    u.unite(u.find(3), u.find(0));

    EXPECT_NE(u.find(2), u.find(3));

    u.unite(u.find(2), u.find(0));

    EXPECT_EQ(u.find(2), u.find(3));
    EXPECT_NE(u.find(2), u.find(1));
    EXPECT_NE(u.find(3), u.find(1));

    u.unite(u.find(0), u.find(1));

    EXPECT_EQ(u.find(2), u.find(1));
    EXPECT_EQ(u.find(3), u.find(1));
    EXPECT_EQ(u.find(2), u.find(0));
    EXPECT_EQ(u.find(3), u.find(0));
}

File: graphidx/julia/test/test_unionfind.jl
include("../src/utils/unionfind.jl")

using Test


@testset "UnionFind, mini_test_2         " begin
    u = UnionFind(2)
    @test find(u, 1) == 1
    @test find(u, 2) == 2

    unite!(u, find(u, 1), find(u, 1))
    @test find(u, 1) == 1
    @test find(u, 2) == 2

    unite!(u, find(u, 1), find(u, 2))
    @test find(u, 1) == find(u, 2)
end


@testset "UnionFind, mini_test_3         " begin
    u = UnionFind(3)
    @test find(u, 1) == 1
    @test find(u, 2) == 2
    @test find(u, 3) == 3

    unite!(u, find(u, 3), find(u, 1))

    @test find(u, 3) == find(u, 1)
    @test find(u, 2) != find(u, 1)
    @test find(u, 2) != find(u, 3)

    unite!(u, find(u, 2), find(u, 1))
    @test length(Set(find(u, i) for i = 1:3)) == 1
end


@testset "UnionFind, mini_test_4         " begin
    u = UnionFind(4)
    for i = 1:4
        @test find(u, i) == i
    end

    unite!(u, find(u, 4), find(u, 1))

    @test find(u, 4) == find(u, 1)
    @test find(u, 3) != find(u, 4)
    @test find(u, 3) != find(u, 1)

    unite!(u, find(u, 4), find(u, 3))

    @test find(u, 4) == find(u, 1)
    @test find(u, 3) == find(u, 4)
    @test find(u, 3) == find(u, 1)
end
